[MODGOBI-146] - Logging improvement

azizbekxm opened the pull request:
<!--
  If you have a relevant JIRA issue number, please put it in the issue title.
  Example: MODGOBI-43 Move "Renewals" information to Purchase Order level

  TL;DR
    - https://www.youtube.com/watch?v=5aHmO_S8FQ4
    - http://www.olitreadwell.com/2016/05/22/how-to-write-great-pull-requests/
    - https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests
-->

## Purpose
<!--
  Why are you making this change? There is nothing more important
  to provide to the reviewer and to future readers than the cause
  that gave rise to this pull request. Be careful to avoid circular
  statements like "the purpose is to update the schema." and
  instead provide an explanation like "there is more data to be provided and stored for Purchase Orders 
  which is currently missing in the schema"

  The purpose may seem self-evident to you now, but the standard to
  hold yourself to should be "can a developer parachuting into this
  project reconstruct the necessary context merely by reading this
  section."

  If you have a relevant JIRA issue, add a link directly to the issue URL here.
  Example: https://issues.folio.org/browse/MODGOBI-43
 -->
https://issues.folio.org/browse/MODGOBI-146
## Approach
<!--
 How does this change fulfill the purpose? It's best to talk
 high-level strategy and avoid code-splaining the commit history.

 The goal is not only to explain what you did, but help other
 developers *work* with your solution in the future.
-->
Logging has been improved for more productive troubleshooting based on documentation https://wiki.folio.org/display/DD/Logging 

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
- [ ] Check logging.
-->

## Learning
<!-- OPTIONAL
  Help out not only your reviewer, but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or addons used
  to solve this problem.
-->

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code are 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added, or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
- Did the task completed based on requirements
 - [ ] Log message must be informative and contain business parameters/indicators
 - [ ] Logging of import business logic actions and steps
 - [ ] Logging of any edge cases scenarios and essential conditional business logic paths
 - [ ] Logging of all exceptional cases
 - [ ] Avoid logging sensitive information or use a specific method of Folio Spring Base to produce messages
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally, all of the PRs involved in breaking changes would be merged on the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of inter-module and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

azizbekxm commented:
it is a partially completed the task, if it needs more improvement, please ping me

azizbekxm commented on src/main/java/org/folio/rest/core/RestClient.java:
```diff
@@ -71,7 +70,9 @@ public Future<JsonObject> handleGetRequest(String endpoint) {
     if (logger.isDebugEnabled()) {
-      logger.debug("Sending 'PUT {}' with body: {}", endpoint, recordData.encodePrettily());
+      logger.debug("Trying to update {} object with data: {}", endpoint, recordData.encodePrettily());
+//    }
```
which type of pattern is best 
old logging pattern 
-  **sending HTTP.METHOD with body { }**

new logging pattern based on documentation 
- **Trying to verb [method] {} object with data: {}**
![image](https://user-images.githubusercontent.com/113523904/201312488-b6b71801-28a9-4910-983c-85f396f918fb.png)

azizbekxm replied:
Should I leave sending pattern or should I change to trying to verb..... pattern

SerhiiNosko replied:
This pattern looks better as for me:
logger.debug("Trying to update {} object with data: {}", endpoint, recordData.encodePrettily());

azizbekxm commented on src/main/java/org/folio/rest/impl/BaseApi.java:
```diff
@@ -41,12 +41,14 @@ public Response buildNoContentResponse() {
         .entity(body)
         .build();
     } catch (URISyntaxException e) {
+      logger.error("Failed to build response with location", e);
```
Should I use method name beginning of message
`logger.error("buildResponseWithLocation::: Failed to build response with location", e);`

based on documentation they mostly don't use
![image](https://user-images.githubusercontent.com/113523904/201314097-c7e3f54b-ec55-4e13-91f6-5ea6a961fcfb.png)

in warn message they use
![image](https://user-images.githubusercontent.com/113523904/201314160-ddc40987-dedf-45ba-a50d-37687ad67c86.png)

SerhiiNosko replied:
For error level logs I think we can omit prefix with method name

azizbekxm commented:
Should I add logging inGobi Custom Mapping Serivce, This class looks small and simple, So I can't decide whether I should add or not
https://github.com/folio-org/mod-gobi/blob/e80306e6c22400b2f649e5601b0e7a6068fc092f/src/main/java/org/folio/rest/service/GobiCustomMappingsService.java

SerhiiNosko commented on src/main/java/org/folio/rest/core/RestClient.java:
```diff
@@ -89,7 +91,13 @@ public Future<JsonObject> post(String endpoint, JsonObject recordData) {
+//      CALLING_ENDPOINT_MSG = "Sending {} {}";
+//      logger.debug(CALLING_ENDPOINT_MSG, HttpMethod.DELETE, endpointById);
+      logger.debug("Tying to delete with using endpoint: {}", endpointById);
+      logger.debug("Sending 'DELETE {}'", endpointById);
```
logger.debug("Tying to delete with using endpoint: {}", endpointById); - this is preferable

azizbekxm commented:
- reverted test/resource/log4j2.properties
- reverted main/resource/log4j2.properties
- fixed main/resource/log4j2.properties with new configuration 
- fixed previous issues

Abdulkhakimov commented on src/main/java/org/folio/gobi/LookupService.java:
```diff
@@ -74,19 +74,21 @@ public LookupService(RestClient restClient) {
+          logger.warn("lookupLocationId:: Location '{}' not found", location);
           return completedFuture(null);
         }
+        logger.info("lookupLocationId:: found location id: {}", locationId);
```
found location **by** id

Abdulkhakimov commented on src/main/java/org/folio/rest/core/RestClient.java:
```diff
@@ -57,7 +56,10 @@ public String getTenantId() {
   public Future<JsonObject> handleGetRequest(String endpoint) {
-    logger.debug("Calling GET {}", endpoint);
+    if (logger.isDebugEnabled()) {
+      logger.debug("Trying to get '{}' object", endpoint);
```
I don't think that we need there logger.isDebugEnabled()

Abdulkhakimov commented on src/main/java/org/folio/rest/core/RestClient.java:
```diff
@@ -89,7 +92,10 @@ public Future<JsonObject> post(String endpoint, JsonObject recordData) {
   public Future<Void> delete(String endpointById) {
-    logger.debug(CALLING_ENDPOINT_MSG, HttpMethod.DELETE, endpointById);
+    if (logger.isDebugEnabled()) {
+      logger.debug("Tying to delete with using endpoint: {}", endpointById);
```
Also there

Abdulkhakimov commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -92,6 +92,7 @@ public CompletableFuture<CompositePurchaseOrder> mapToPurchaseOrder(Document doc
 
 
   public static OrderMappings.OrderType getOrderType(Document doc) {
+    logger.debug("getOrderType:: Trying to get order type from '{}'", doc.getDoctype());
```
Could you add checking to debug is enabled?

Abdulkhakimov commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -120,7 +122,7 @@ public CompletableFuture<Document> parse(String entity) {
       return CompletableFuture.completedFuture(parser.parse(entity));
     } catch (GobiPurchaseOrderParserException e) {
-      logger.error("Failed to parse GobiPurchaseOrder", e);
+      logger.error("Failed to parse GobiPurchaseOrder: {}", entity, e);
```
Please add carry over to the next line!

Abdulkhakimov commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -209,9 +211,10 @@ private CompletableFuture<String> placeOrder(CompositePurchaseOrder compPO) {
 
 
   public CompletableFuture<String> getOrPlaceOrder(CompositePurchaseOrder compPO) {
+    logger.debug("getOrPlaceOrder:: Trying to get order or place it with composite PO: {}", Json.encodePrettily(compPO));
```
Also there we need checking to debug is enabled.

Abdulkhakimov commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -255,26 +259,27 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
       })
       .exceptionally(t -> {
-        logger.error("Exception looking up for existing Order", t);
+        logger.error("Error when looking up for existing Order", t);
```
please add to logs Composite PO line id

Abdulkhakimov commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -255,26 +259,27 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
 
   private CompletableFuture<CompositePurchaseOrder> getExistingOrderById(CompositePurchaseOrder compositePurchaseOrder) {
-    logger.info("Retrieving existing Order with ID {}", compositePurchaseOrder.getId());
+    logger.debug("getExistingOrderById:: Trying to retrieve existing Order with ID: {}", compositePurchaseOrder.getId());
```
is debug enabled

SerhiiNosko commented on src/main/java/org/folio/gobi/LookupService.java:
```diff
@@ -259,6 +270,7 @@ public CompletableFuture<String> lookupConfigAddress(String shipToName) {
+        logger.warn("lookupConfigAddress:: Config address with name '{}' not found", shipToName);
         return null;
       })
       .exceptionally(t -> {
```
Please add shipToName to this error log

SerhiiNosko commented on src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:
```diff
@@ -120,7 +122,7 @@ public CompletableFuture<Document> parse(String entity) {
       return CompletableFuture.completedFuture(parser.parse(entity));
     } catch (GobiPurchaseOrderParserException e) {
-      logger.error("Failed to parse GobiPurchaseOrder", e);
+      logger.error("Failed to parse GobiPurchaseOrder: \n {}", entity, e);
```
Why we need \n here?

SerhiiNosko commented on src/main/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImpl.java:
```diff
@@ -37,10 +37,10 @@ public void postGobiOrders(String entity, Map<String, String> okapiHeaders,
     PostGobiOrdersHelper helper = new PostGobiOrdersHelper(asyncResultHandler, okapiHeaders, vertxContext);
 
-    logger.info("Parsing Request: \n {}", entity);
+    logger.info("postGobiOrders:: Trying to parse request: \n {}", entity);
```
please delete '\n' from logs

SerhiiNosko commented on src/main/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImpl.java:
```diff
@@ -37,10 +37,10 @@ public void postGobiOrders(String entity, Map<String, String> okapiHeaders,
     helper.parse(entity)
       .thenCompose(gobiPO -> {
-        logger.info("Mapping Request...");
+        logger.info("postGobiOrders:: Trying to create PO from request: \n {}", gobiPO);
```
please delete '\n' from logs
